### PR TITLE
RFC: Enabling oneDNN operations for x86 CPUs

### PR DESCRIPTION
**Pull request by penpornk**
This RFC will be open for comment until Tuesday, November 9th, 2021.

| Status        | Proposed                                             |
:-------------- |:---------------------------------------------------- |
| **RFC #**     | [400](https://github.com/tensorflow/community/pull/400)                                                 |
| **Author(s)** | Penporn Koanantakool (penporn@google.com)            |
| **Sponsor**   | Tatiana Shpeisman (shpeisman@google.com)             |
| **Updated**   | 2021-10-25                                           |

# Objective
Default TensorFlow Linux packages can efficiently utilize CPU hardware features. Make it easier to support new features.

### CODE REVIEW

File: rfcs/20210930-enable-onednn-ops.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+## Design Proposal
+
+The design consists of three main components: oneDNN custom ops, graph rewrite passes, and a software-managed cache.
+
+**oneDNN custom ops** are TensorFlow ops that perform some set up (e.g., creating oneDNN compute primitives, memory descriptors, etc) before calling oneDNN routines.
+* **Visibility:** These custom ops are hidden from users in the Python layer, cannot be saved in SavedModels, and are only introduced into the TensorFlow graph by graph rewrite passes in the C++ layer.
+* **Scope and signature:** Most ops have a 1:1 mapping to a standard TensorFlow op, with the same op signatures (inputs, attributes, outputs, etc). The rest are fused operations (one op is equivalent to a sequence of standard TensorFlow ops).
+* **Expected tensor format:** oneDNN ops process tensors directly in TensorFlow’s native format (NHWC), e.g., no tensor layout conversions.
```
Comment (Jianhui-Li) [inline rfcs/20210930-enable-onednn-ops.md:46]: I understand that for activation, we want to hide the blocked layout completely.  But for weight, I believe that framework may want to cache the reordered weights for the inference task.  Can you elaborate the expectation for weight layout?

**Comment by penpornk** [comment]
We have tentatively scheduled a public design review meeting for this RFC.
**Time:** Thursday, November 11th, 2021 from 3-4pm PT
**Meeting link:** [Google Meet](https://meet.google.com/nky-rahk-xxv)

Meeting participants are expected to read the RFC ahead of time as the meeting will focus on the remaining issues/questions.

If there are no outstanding issues on this RFC by the time the feedback phase closes on Tuesday November 9th, 2021, we may cancel the design review meeting. (We will update here.)

**Comment by penpornk** [comment]
The design review meeting has been canceled since there are no outstanding issues to discuss.

**Review by ematejska — APPROVED** [review]

**Comment by penpornk** [comment]
A quick follow-up that [oneDNN operations](https://github.com/tensorflow/community/blob/master/rfcs/20210930-enable-onednn-ops.md) were enabled by default on [recent x86 CPUs on Linux starting from TF 2.9](https://blog.tensorflow.org/2022/05/whats-new-in-tensorflow-29.html). 
Next, we plan to also enable them by default on:
* aarch64 CPUs on Linux (starting with Neoverse V1) 
* x86 CPUs on Windows

If anyone has any concerns, please post them here by **July 12, 2023**.

Thank you!

**Pull request merged**